### PR TITLE
Refactored/Decoupled Input string generation from the core logic of interpreter

### PR DESCRIPTION
**Pull request by nikasakandelidze**
Abstract ( Ideas behind why certain things were changed ):
Although code worked perfectly fine, the coupling between different components was high( it still might be a bit high between some of them ). What i especially  noticed was that the core logic of interpreter ( "lox" file ) had code of getting input ( from file or from cli, in the future potentially from other sources ) injected right into itself.
This specific design decision IMHO might be problematic because of 2 reasons:
- It's harder to understand code since all the different components/logics are put together
- If i want to introduce some change in input generation code i have to still open and go through  core logic file.
- Later if i want to add new input source ( like making interpreter work via network using HTTP/REST ) ill have to update to work with core logic components which is not desirable.

In the best case i want to work with only 1 component/file in case of:
- introducing updates to existing input generation
- added new input source

What was changed:
- The main part of refactoring was to introduce separate module for generating input sources and remove input generation ( via file or via cli ) from the core logic.
- Lox core module should simply get input string and shouldn't have responsibility or knowledge of where this input came from.
- The logic/decision of which Input source: either cli or  file to use ( depending on input arguments ) shouldn't be made by core interpreter logic. This decision is for the main application flow.

I might have messed something up so please share your thoughts.
ty.

### CODE REVIEW

File: src/InputSource/index.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
+import fs from "fs";
+import reader from "readline-sync";
+
+import { Optional } from "../util/types";
+
+export type InputSourceType = "file" | "cli";
+
+export interface InputSource {
+  getInput(): Optional<string>;
```
Comment (dudupopkhadze) [inline src/InputSource/index.ts]: You can just to `getInput(): string | undefined` here, no need for an extra type

File: src/InputSource/index.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+export class InputSourceFactory {
```
Comment (dudupopkhadze) [inline src/InputSource/index.ts:10]: nice

File: src/InputSource/index.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+class FileInputSource implements InputSource {
+  private source: string;
+
+  constructor(source: string) {
+    this.source = source;
+  }
+
+  getInput(): Optional<string> {
```
Comment (dudupopkhadze) [inline src/InputSource/index.ts]: I don't think you need to specify return type for `getInput`implementations, most likely TS will infer them automatically

File: src/InputSource/index.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    } catch (error) {
+      console.log(error);
+      console.log(`Unable to read file: ${this.source}`);
+      return undefined;
+    }
+  }
+}
+
+class CliInputSource implements InputSource {
+  getInput(): Optional<string> {
```
Comment (dudupopkhadze) [inline src/InputSource/index.ts]: same here

File: src/core/lox.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
+import { Interpreter } from "./Interpreter";
+import { Parser } from "./Parser";
+import { RuntimeError } from "./RuntimeError";
+import { Scanner } from "./Scanner";
+import { Token, TokenType } from "./Token";
+import { Resolver } from "./Resolver";
+
+export class Lox {
```
Comment (dudupopkhadze) [inline src/core/lox.ts:8]: Yep much cleaner this way 👍

**Review by dudupopkhadze — APPROVED** [review]
Really good work here, Thanks !

Logix of the interpreter was indeed too intertwined with logic of getting input source, now it's way more customisable.

Also now i think that using Visitor Pattern so much was kinda overkill , since it introduced a lot of boilerplate that could have been avoided.

**Comment by nikasakandelidze** [comment]
Updated PR by removing excess custom type usage.
P.S. i can't merge pr i guess it's your privilege:) so please do it whenever you feel like it.

**Comment by dudupopkhadze** [comment]
Thanks for the contribution !

**Pull request merged**